shell/platform/darwin/ios/framework/Source/FlutterEngine.mm

```diff
@@ -171,6 +171,7 @@ - (void)setViewController:(FlutterViewController*)viewController {
 }
 
 - (void)notifyViewControllerDeallocated {
+  [[self lifecycleChannel] sendMessage:@"AppLifecycleState.detached"];
```

Comment:
I don't know how to test this.

Reply:
@xster has been working on some tests related to these and can probably help

Reply:
I think https://github.com/flutter/engine/pull/11890 might make some part of your test a bit easier. You'd have to change add a bit to push another VC after the Flutter VC and change the assert value after you pop the Flutter VC. 

I expect to have that one reworked (retranslate to obj-c) by monday

Reply:
Updated existing test

Reply:
I'm tempted to suggest that we keep the logic that does all this stuff in the same place. i.e. all the rest of the lifecycle system channel transitions are sourced in flutterviewcontroller (as is the source of the signal that triggers this call). We might as well keep them all in the same place.

Reply:
When flutterviewcontroller is destroyed it does not know if it is still attached to engine.
For example,
In the beginning, fluttervc1 attached to engine.
At some point, fluttervc2 replace fluttervc1 and attaches to engine.
We don't have the logic(and I don't know if we should) to clean out fluttervc1's engine instance. When fluttervc1 later on gets dealloc, it does not know whether it should make the engine go to detached or not.

That is why I do it in engine side, because the engine knows when it should go to detached. If we want to group the life cycle event management into fluttervc, fluttervc has to know the status of engine, whether the engine is attached to fluttervc or not.

Reply:
umm that's a good point. I think that's also generally true for all other references to the engine from the VC in the first place. 

Maybe there isn't any immediate, great solutions. I'm tempted to say, since there's already ~40 "incorrect" _engine.get() in the flutter vc, we might as well keep this one in the same place (for easier refactor in the future since they're all together) and leave a bug for us to clean this up together when the engine sets a flutter view controller that's not the current one. 

WDYT @dnfield, @gaaclarke?

Reply:
For this specific one, this seems like the right move. If I understand correctly, this is specifically to let us know about a detatched state of the VC - the VC can't necessarily tell us that because it might try to tell us too late (at a point where it's reference to the engine is lost).

It would be worth auditing other places where we try to send information via the engine, but as long as it's information that only a makes sense in an attached state it should be fine to keep it in the VC.  If not, we should move it to the engine.

Reply:
For instance, there are a few lifecycle messages we send based on whether the view is coming or going about the Application state - those probably should get refactored so they work more sensibly in add-to-app contexts if that's possible.

The other usages of `_engine.get()` all look like they are correct to me.

Reply:
I will leave this as is.

Reply:
tangential to this PR now: 

> as long as it's information that only a makes sense in an attached state it should be fine to keep it in the VC

the point @chunhtai was pointing to originally I think is that if the VC did get detached from the engine, the VC would never know and would crash at some point. It's not happening right now because removing the engine from a currently active foreground vc is not something people do commonly. But we should audit.